### PR TITLE
feat: add `EffectBinder` phase

### PR DESCRIPTION
**Pull request by JonathanStarup**

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  private def letBindEffectsStmt(stmt: Stmt)(implicit flix: Flix): Stmt = stmt match {
+    case Stmt.Ret(expr, tpe, loc) => Stmt.Ret(letBindEffectsTopLevel(expr), tpe, loc)
+  }
+
+  /**
+    * Identity Function.
+    */
+  private def letBindEffectsTopLevel(exp: Expr)(implicit flix: Flix): Expr = exp match {
+    // simple
+    case _: Expr.Cst | _: Expr.Var | _: Expr.ApplyAtomic | _: Expr.ApplyClo | _: Expr.ApplyDef |
```
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala:60]: This is really just a question of maintaining the same style.

File: main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+import scala.annotation.tailrec
+import scala.collection.mutable
+
+
+object EffectBinder {
+
+  /**
+    * Identity Function.
+    */
+  def run(root: Root)(implicit flix: Flix): Root = flix.phase("EffectBinder") {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala:35]: Can we come up with a better name than EffectBinder? I guess its not (yet) FG CBV?

File: main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.Ast.BoundBy
+import ca.uwaterloo.flix.language.ast.ReducedAst._
+import ca.uwaterloo.flix.language.ast._
+import ca.uwaterloo.flix.util.collection.Chain
+import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps}
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala:28]: extra space

File: main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+object EffectBinder {
+
+  /**
+    * Identity Function.
+    */
+  def run(root: Root)(implicit flix: Flix): Root = flix.phase("EffectBinder") {
+    val newDefs = ParOps.parMapValues(root.defs)(letBindEffectsDef)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala:36]: No local ctx yet?

File: main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    // fancy
+    case Expr.Branch(exp, branches, tpe, purity, loc) =>
+      val e = letBindEffectsTopLevel(exp)
+      val branches1 = branches.map {
+        case (sym, branchExp) => (sym, letBindEffectsTopLevel(branchExp))
+      }
+      Expr.Branch(e, branches1, tpe, purity, loc)
+    case Expr.JumpTo(sym, tpe, purity, loc) =>
+      Expr.JumpTo(sym, tpe, purity, loc)
+    case Expr.Let(sym, exp1, exp2, tpe, purity, loc) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala:73]: I would add line breaks between each case, to get some "air"

File: main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala (hunk 2 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala:35]: StacklessEffects?

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
Will merge soon.

**Comment by magnus-madsen** [comment]
@JonathanStarup Merged!

**Pull request merged**